File: 2020/GTBIT/L001_foundation/l006_Recursion/l002.cpp
#include <iostream>
#include <vector>

using namespace std;

void printSS(string str, int idx, string ans)
{
    if (idx == str.length())
    {
        cout << ans << endl;
        return;
    }

    printSS(str, idx + 1, ans + str[idx]);
    printSS(str, idx + 1, ans);
}

vector<string> subseq_03(string str, int idx)
{
    if (idx == str.length())
    {
        vector<string> base;
        base.push_back("");
        return base;
    }

    vector<string> recAns = subseq_03(str, idx + 1);

    vector<string> myAns(recAns); // nahi anne ki choise dekhli.

    char ch = str[idx];
    for (string s : recAns)
    {
        myAns.push_back(ch + s);
    }

    return myAns;
}

// sr - source row
// sc - source column
// dr - destination row
// dc - destination column
vector<string> getMazePaths(int sr, int sc, int dr, int dc)
{
    if (sr == dr && sc == dc)
    {
        vector<string> base;
        base.push_back("");
        return base;
    }

    vector<string> myAns;
    if (sc + 1 <= dc)
    {
        vector<string> Horizontal = getMazePaths(sr, sc + 1, dr, dc);
        for (string s : Horizontal)
        {
            myAns.push_back("h" + s);
        }
    }

    if (sr + 1 <= dr)
    {
        vector<string> Vertical = getMazePaths(sr + 1, sc, dr, dc);
        for (string s : Vertical)
        {
            myAns.push_back("v" + s);
        }
    }

    return myAns;
}

// sr - source row
// sc - source column
// dr - destination row
// dc - destination column
vector<string> getMazePathsDiagonal(int sr, int sc, int dr, int dc)
{
    if (sr == dr && sc == dc)
    {
        vector<string> base;
        base.push_back("");
        return base;
    }

    vector<string> myAns;
    if (sc + 1 <= dc)
    {
        vector<string> Horizontal = getMazePathsDiagonal(sr, sc + 1, dr, dc);
        for (string s : Horizontal)
        {
            myAns.push_back("h" + s);
        }
    }

    if (sr + 1 <= dr)
    {
        vector<string> Vertical = getMazePathsDiagonal(sr + 1, sc, dr, dc);
        for (string s : Vertical)
        {
            myAns.push_back("v" + s);
        }
    }

    if (sr + 1 <= dr && sc + 1 <= dc)
    {
        vector<string> Diagonal = getMazePathsDiagonal(sr + 1, sc + 1, dr, dc);
        for (string s : Diagonal)
        {
            myAns.push_back("d" + s);
        }
    }

    return myAns;
}

int getMazePaths(int sr, int sc, int dr, int dc, string ans)
{
    if (sr == dr && sc == dc)
    {
        cout << ans << endl;
        return 1;
    }

    int count = 0;
    if (sc + 1 <= dc)
        count += getMazePaths(sr, sc + 1, dr, dc, ans + 'h');
    if (sr + 1 <= dr)
        count += getMazePaths(sr + 1, sc, dr, dc, ans + 'v');

    return count;
}

vector<string> getMazePaths_multiMoves(int sr, int sc, int dr, int dc)
{
    if (sr == dr && sc == dc)
    {
        vector<string> base;
        base.push_back("");
        return base;
    }

    vector<string> myAns;
    for (int jump = 1; sc + jump <= dc; jump++)
    {
        vector<string> Horizontal = getMazePaths_multiMoves(sr, sc + jump, dr, dc);
        for (string s : Horizontal)
        {
            myAns.push_back("h" + to_string(jump) + s);
        }
    }

    for (int jump = 1; sr + jump <= dr; jump++)
    {
        vector<string> Vertical = getMazePaths_multiMoves(sr + jump, sc, dr, dc);
        for (string s : Vertical)
        {
            myAns.push_back("v" + to_string(jump) + s);
        }
    }

    for (int jump = 1; sr + jump <= dr && sc + jump <= dc; jump++)
    {
        vector<string> Digonal = getMazePaths_multiMoves(sr + jump, sc + jump, dr, dc);
        for (string s : Digonal)
        {
            myAns.push_back("d" + to_string(jump) + s);
        }
    }

    return myAns;
}

void getMazePaths_multiMoves(int sr, int sc, int dr, int dc, string ans)
{
    if (sr == dr && sc == dc)
    {
        cout << ans << endl;
        return;
    }

    for (int jump = 1; sc + jump <= dc; jump++)
        getMazePaths_multiMoves(sr, sc + jump, dr, dc, ans + 'h' + to_string(jump));
    for (int jump = 1; sr + jump <= dr; jump++)
        getMazePaths_multiMoves(sr + jump, sc, dr, dc, ans + 'v' + to_string(jump));
    for (int jump = 1; sr + jump <= dr && sc + jump <= dc; jump++)
        getMazePaths_multiMoves(sr + jump, sc + jump, dr, dc, ans + 'd' + to_string(jump));
}

int main()
{
    string str = "abc";
    printSS(str, 0, "");

    vector<string> ans = subseq_03(str, 0);
    for (string s : ans)
        cout << s << " ";

    return 0;
}

File: 2020/GTBIT/L001_foundation/l006_Recursion/l002.java
import java.util.ArrayList;

public class l002 {

    public static void printSS(String str, int idx, String ans) {
        if (idx == str.length()) {
            System.out.println(ans);
            return;
        }

        printSS(str, idx + 1, ans + str.charAt(idx));
        printSS(str, idx + 1, ans);
    }

    public static void printSS_02(String str, int idx, StringBuilder ans) {
        if (idx == str.length()) {
            System.out.println(ans);
            return;
        }

        ans.append(str.charAt(idx));
        printSS_02(str, idx + 1, ans);
        ans.deleteCharAt(ans.length() - 1);

        printSS_02(str, idx + 1, ans);
    }

    // return Type
    public static ArrayList<String> subseq_03(String str, int idx) {
        if (idx == str.length()) {
            ArrayList<String> base = new ArrayList<>();
            base.add("");
            return base;
        }

        ArrayList<String> recAns = subseq_03(str, idx + 1);

        ArrayList<String> myAns = new ArrayList<>(recAns); // nahi anne ki choise dekhli.

        char ch = str.charAt(idx);
        for (String s : recAns) {
            myAns.add(ch + s);
        }

        return myAns;
    }

    public static void printASCIISS(String str, int idx, String ans) {
        if (idx == str.length()) {
            System.out.println(ans);
            return;
        }

        printASCIISS(str, idx + 1, ans + str.charAt(idx));
        printASCIISS(str, idx + 1, ans + (int) str.charAt(idx));
        printASCIISS(str, idx + 1, ans);
    }

    public static ArrayList<String> printASCIISS_Ret(String str, int idx) {
        if (idx == str.length()) {
            ArrayList<String> base = new ArrayList<>();
            base.add("");
            return base;
        }

        ArrayList<String> recAns = printASCIISS_Ret(str, idx + 1);

        ArrayList<String> myAns = new ArrayList<>(); // nahi anne ki choise dekhli.

        char ch = str.charAt(idx);
        for (String s : recAns) {
            myAns.add(ch + s);
            myAns.add((int) ch + s);
            myAns.add(s);
        }

        return myAns;
    }

    public static void subsequence() {
        String str = "abc";
        // printSS(str, 0, "");
        // printSS_02(str,0,new StringBuilder());
        // System.out.println(subseq_03(str, 0));
        printASCIISS(str, 0, "");
    }

    static String[] codes = { ".;", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tu", "vwx", "yz" };

    public static ArrayList<String> getKPC(String str, int idx) {
        if (idx == str.length()) {
            ArrayList<String> base = new ArrayList<>();
            base.add("");
            return base;
        }

        ArrayList<String> recAns = getKPC(str, idx + 1);
        ArrayList<String> myAns = new ArrayList<>();

        char ch = str.charAt(idx);
        String code = codes[ch - '0'];
        int count = 0;

        for (int i = 0; i < code.length(); i++) {
            char c = code.charAt(i);

            for (String s : recAns) {
                myAns.add(c + s);
            }
        }

        return myAns;
    }

    public static int printKPC(String str, int idx, String ans) {
        if (idx == str.length()) {
            System.out.println(ans);
            return 1;
        }

        char ch = str.charAt(idx);
        String code = codes[ch - '0'];
        int count = 0;

        for (int i = 0; i < code.length(); i++) {
            char c = code.charAt(i);
            count += printKPC(str, idx + 1, ans + c);
        }

        return count;
    }

    public static void getMazePaths_multiMoves(int sr, int sc, int dr, int dc, String ans) {
        if (sr == dr && sc == dc) {
            System.out.println(ans);
            return;
        }

        for (int jump = 1; sc + jump <= dc; jump++)
            getMazePaths_multiMoves(sr, sc + jump, dr, dc, ans + 'h' + jump);
        for (int jump = 1; sr + jump <= dr; jump++)
            getMazePaths_multiMoves(sr + jump, sc, dr, dc, ans + 'v' + jump);
        for (int jump = 1; sr + jump <= dr && sc + jump <= dc; jump++)
            getMazePaths_multiMoves(sr + jump, sc + jump, dr, dc, ans + 'd' + jump);

    }

    public static ArrayList<String> getMazePaths_multiMoves(int sr, int sc, int dr, int dc) {

        if (sr == dr && sc == dc) {
            ArrayList<String> base = new ArrayList<>();
            base.add("");
            return base;
        }

        ArrayList<String> myAns = new ArrayList<>();
        for (int jump = 1; sc + jump <= dc; jump++) {
            ArrayList<String> Horizontal = getMazePaths_multiMoves(sr, sc + jump, dr, dc);
            for (String s : Horizontal) {
                myAns.add("h" + jump + s);
            }
        }

        for (int jump = 1; sr + jump <= dr; jump++) {
            ArrayList<String> Vertical = getMazePaths_multiMoves(sr + jump, sc, dr, dc);
            for (String s : Vertical) {
                myAns.add("v" + jump + s);
            }
        }

        for (int jump = 1; sr + jump <= dr && sc + jump <= dc; jump++) {
            ArrayList<String> Digonal = getMazePaths_multiMoves(sr + jump, sc + jump, dr, dc);
            for (String s : Digonal) {
                myAns.add("d" + jump + s);
            }
        }

        return myAns;
    }

    public static void main(String[] args) {
        subsequence();
    }

}